### PR TITLE
[App] Add set auto suggestion feature

### PR DESCRIPTION
**Pull request by hellozepp**
1. Fix save does not prompt option

### CODE REVIEW

File: external/mlsql-autosuggest/src/main/java/tech/mlsql/autosuggest/AutoSuggestContext.scala [hunk 1]
```diff
@@ -129,7 +129,7 @@ class AutoSuggestContext(val session: SparkSession,
   }
 
   def toRelativePos(tokenPos: TokenPos): (TokenPos, Int) = {
-    var skipSize = 0
+    var skipSize = -1
```
Comment (allwefantasy) [inline external/mlsql-autosuggest/src/main/java/tech/mlsql/autosuggest/AutoSuggestContext.scala]: Can you explain why we need to change skipSize to -1?

File: external/mlsql-autosuggest/src/main/java/tech/mlsql/autosuggest/statement/SetSuggester.scala [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+  override def suggest(): List[SuggestItem] = {
+    keywordSuggest ++ defaultSuggest(subSuggesters.toMap)
+  }
+
+
+  private def keywordSuggest: List[SuggestItem] = {
+    _tokenPos match {
+      case TokenPos(pos, TokenPosType.NEXT, offsetInToken) =>
+        var items = List[SuggestItem]()
+        val temp = TokenMatcher(_tokens, pos).back.
```
Comment (allwefantasy) [inline external/mlsql-autosuggest/src/main/java/tech/mlsql/autosuggest/statement/SetSuggester.scala]: Will statement like  `set a="b" ` where the cursor in last pos triggers the where keyword completion?

**Review by allwefantasy — CHANGES_REQUESTED** [review]
Please provide some unit tests.

File: external/mlsql-autosuggest/src/main/java/tech/mlsql/autosuggest/AutoSuggestContext.scala (hunk 1 above)
Comment (hellozepp) [reply external/mlsql-autosuggest/src/main/java/tech/mlsql/autosuggest/AutoSuggestContext.scala]: Wrong modification, has been restored

File: external/mlsql-autosuggest/src/main/java/tech/mlsql/autosuggest/statement/SetSuggester.scala (hunk 2 above)
Comment (hellozepp) [reply external/mlsql-autosuggest/src/main/java/tech/mlsql/autosuggest/statement/SetSuggester.scala]: ---------------
in Chinese
---------------

# 背景

目前，我们支持 load/select/save 三个语法，但还不具备**Production Ready**。需要把其他语法的代码提示补全。添加顺序按照： set 语法， run/train 语法， regsiter 语法。



## Set 基础应用

下面列举set语法的用法，后面会根据使用方式设计具体的语法提示：

```
-- set字符串变量

set hello='world';

-- 或者使用双引号。单引号和双引号都可以引用变量

set hello="world";

select "hello ${hello}" as title as output;



-- set文本内容,文本中存在引号等特殊字符可以正常使用

set rawData=''' 

{"jack":1,"jack2":2}

{"jack":2,"jack2":3}

''';





-- 密码为特殊符号

set rawData="#[['/*.?]]";





-- set表名

set hello="world";

select "hello William" as title 

as `${hello}`;



-- type = conf表示这是一个配置选项，通常用于配置系统的行为，比如：

set spark.sql.shuffle.partitions=200 where type="conf";



-- type = shell,也就是set后的key最后是由shell执行生成的。 我们已经不推荐使用该方式,比如：

set date=`date` where type="shell";

select "${date}" as dt as output;



-- type = sql类型，这意味着set后的key最后是由sql引擎执行生成的。用法：

set date=`select date_sub(CAST(current_timestamp() as DATE), 1) as dt` 

where type="sql";

select "${date}" as dt as output;



-- defaultParam类型：

set hello="foo";

set hello="bar" where type="defaultParam";

select "${hello}" as name as output;



-- mode

-- 内置变量
```

## Set 语法提示

按照目前set语法支持的内容，语法提示分为下面2部分：

1. **Option autosuggestion** （where后面可选项提示）

set 的where参数提示 **type**、**mode**

- type
  - text ： 文本内容

- conf :提示系统参数，`set spark.sql.shuffle.partitions=200 where type="conf";`

- shell :提示shell脚本语法，如果where前的字符为单撇号，应该仅提示shell

- sql :提示sql语法

- defaultParam :为默认值

- Mode autosuggestion 
  - `**runtime**`**：**希望set语句可以运行时就evaluate值

- `**compile**`**：**希望set语句可以编译时就evaluate值



注意，如果option的值已经有输入但不全 ，需要有提示补全，有如下几种case：

1. string为封闭的，其前面的字符是Option值的前缀，比如 `type="c[cursor]"`=>  `Option + Space` hit => `type="c[onf]"`
2. string为封闭的，其前面的字符不是Option值的前缀，比如 `type="xx[cursor]"`=>  `Option + Space` hit => `type="c[]"`
3. string为非封闭的，其前面的字符是Option值的前缀，比如 `type="c[cursor]`=>  `Option + Space` hit => `type="c[onf"]`
4. string为非封闭的，其前面的字符不是Option值的前缀，比如 `type="xx[cursor]"`=>  `Option + Space` hit => `type="c[]`


---------------
in English
---------------


# background

Currently, we support the three syntaxes of load/select/save, but we don’t have **Production Ready** yet. Need to complete other syntax code Suggestion. The order of addition is as follows: set syntax, run/train syntax, regsiter syntax.


## Set Basic Application

The usage of set grammar is listed below, and specific grammar hints will be designed according to the usage method:

```
- set string variable

set hello='world';

- Or use double quotes. Both single and double quotes can quote variables

set hello="world";

select "hello ${hello}" as title as output;



- Set text content, special characters such as quotation marks in the text can be used normally

set rawData='''

{"jack":1,"jack2":2}

{"jack":2,"jack2":3}

''';





- The password is a special symbol

set rawData="#[['/*.?]]";





- set table name

set hello="world";

select "hello William" as title

as `${hello}`;



- type = conf means this is a configuration option, usually used to configure the behavior of the system, such as:

set spark.sql.shuffle.partitions=200 where type="conf";



- type = shell, that is, the key after set is finally generated by the shell execution. We no longer recommend using this method, such as:

set date=`date` where type="shell";

select "${date}" as dt as output;



- type = sql type, which means that the key after set is finally generated by the sql engine. usage:

set date=`select date_sub(CAST(current_timestamp() as DATE), 1) as dt`

where type="sql";

select "${date}" as dt as output;



- defaultParam type:

set hello="foo";

set hello="bar" where type="defaultParam";

select "${hello}" as name as output;



- mode

- Built-in variables
```

## Set Syntax Tips

According to the content supported by the current set grammar, the grammar prompt is divided into the following two parts:

1. **Option autosuggestion** (optional prompt after where)

Set where parameter prompt **type**, **mode**

-type
  -text: text content

-conf: prompt system parameters, `set spark.sql.shuffle.partitions=200 where type="conf";`

-shell: prompt shell script syntax, if the character before where is a single apostrophe, it should only prompt shell

-sql: prompt sql syntax

-defaultParam: is the default value

-Mode autosuggestion
  -`**runtime**`**: **I want to evaluate the value when the set statement can run

-`**compile**`**: **I hope the set statement can evaluate the value at compile time



Note that if the value of option has been entered but is not complete, it needs to be prompted to complete. There are several cases as follows:

1. The string is closed, and the character before it is the prefix of the Option value, such as `type="c[cursor]"`=> `Option + Space` hit => `type="c[onf]"`
2. The string is closed, and the character before it is not the prefix of the Option value, such as `type="xx[cursor]"`=> `Option + Space` hit => `type="c[]"`
3. The string is non-closed, and the character before it is the prefix of the Option value, such as `type="c[cursor]`=> `Option + Space` hit => `type="c[onf"]`
4. The string is non-closed, and the character before it is not the prefix of the Option value, such as `type="xx[cursor]"`=> `Option + Space` hit => `type="c[]`

File: external/mlsql-autosuggest/src/main/java/tech/mlsql/autosuggest/dsl/TokenMatcher.scala [hunk 3]
```diff
@@ -302,6 +319,18 @@ object TokenMatcher {
     SqlBaseLexer.LIMIT,
     SqlBaseLexer.ORDER
   )
+
+  object SqlStringIdentsEnum extends Enumeration {
+    type SetOptionEnum = Value
+    val STRING: Value = Value(DSLSQLLexer.STRING, "\"")
+    val BLOCK_STRING: Value = Value(DSLSQLLexer.BLOCK_STRING, "'''")
+    val IDENTIFIER: Value = Value(DSLSQLLexer.IDENTIFIER, "\'")
+    val EXECUTE_COMMAND: Value = Value(DSLSQLLexer.EXECUTE_COMMAND, "\"")
```
Comment (allwefantasy) [inline external/mlsql-autosuggest/src/main/java/tech/mlsql/autosuggest/dsl/TokenMatcher.scala:328]: Should the execute command quote  be  " ` " ?

**Review by allwefantasy — APPROVED** [review]

**Review by allwefantasy — APPROVED** [review]

**Pull request merged**